opensourceleg/actuators/base.py

```diff
@@ -0,0 +1,428 @@
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    NamedTuple,
+    Optional,
+    Protocol,
+    Set,
+    TypeVar,
+    Union,
+    cast,
+    runtime_checkable,
+)
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from functools import partial
+
+import numpy as np
+
+from opensourceleg.logging.logger import LOGGER
+
+# TODO: Add validators for every custom data type
+
+
+@dataclass
+class MOTOR_CONSTANTS:
+    MOTOR_COUNT_PER_REV: float
+    NM_PER_AMP: float
+    NM_PER_RAD_TO_K: float
+    NM_S_PER_RAD_TO_B: float
+    MAX_CASE_TEMPERATURE: float
+    MAX_WINDING_TEMPERATURE: float
+
+    def __post_init__(self):
+        if any(x <= 0 for x in self.__dict__.values()):
+            raise ValueError(
+                "All values in MOTOR_CONSTANTS must be non-zero and positive."
+            )
+
+    @property
+    def RAD_PER_COUNT(self) -> float:
+        return 2 * np.pi / self.MOTOR_COUNT_PER_REV
+
+    @property
+    def NM_PER_MILLIAMP(self) -> float:
+        return self.NM_PER_AMP / 1000
+
+
+class CONTROL_MODES(Enum):
```

Comment:
Second thought ... We should not be intended to create CONTROL_MODES enums classes as a default config for users, since most of them are only meaningful for Dephy Actuators (when re-using them representing Specific Registers). A better way for this is a protocol that guides users to inherit the register values provided by the manufacturers, not creating Enum references until at specific modules (e.g. dephy.py, moteus.py)

Still generating ideas about this. Will come back and leave further updates later

Reply:
The `CONTROL_MODES` enum is a global mapping of how various control mode strings map to integers. 

The `CONTROL_MODE_CONFIGS` on the other hand is the container that holds all supported control modes (their max_gains, entry, and exit callbacks). These containers can be indexed using the global `CONTROL_MODES` mapping. 

We want the developers to interact with `CONTROL_MODE_CONFIGS` while still using `CONTROL_MODES` enum to identify which modes are supported. The idea behind having one global enum that maps control mode names to strings is to globally point to control modes regardless of actuator types, which would help a lot during unit testing.

[Code](https://github.com/neurobionics/opensourceleg/blob/d2b1f5b2845f0c5291b4ce94d6bf1441e8e6be60/opensourceleg/actuators/base.py#L215)